pkg/applicationserver/io/grpc/grpc.go

```diff
@@ -213,9 +214,18 @@ func (s *impl) GetMQTTConnectionInfo(ctx context.Context, ids *ttnpb.Application
 var errPayloadCryptoSkipped = errors.DefineFailedPrecondition("payload_crypto_skipped", "payload crypto skipped")
 
 func (s *impl) SimulateUplink(ctx context.Context, up *ttnpb.ApplicationUp) (*emptypb.Empty, error) {
-	if err := rights.RequireApplication(ctx, up.EndDeviceIds.ApplicationIds, ttnpb.Right_RIGHT_APPLICATION_TRAFFIC_UP_WRITE); err != nil {
+	if up.Simulated {
+		// Traffic already marked as simulated is discarded, in order to avoid traffic loops created by injecting
+		// simulated traffic which is then piped into the simulate uplink endpoint itself.
+		return ttnpb.Empty, nil
+	}
```

Comment:
This is an unrelated behavioral change. I don't get the comment. What are we trying to avoid, what would this traffic loop look like? Where is AS calling `SimulateUplink()` itself?

Reply:
The request type for `SimulateUplink` is `ApplicationUp`, and it has the HTTP binding `/as/applications/{end_device_ids.application_ids.application_id}/devices/{end_device_ids.device_id}/up/simulate`. Webhook messages from TTS also have the request body type `ApplicationUp`.

It is possible to create a webhook which uses `/as/applications/{end_device_ids.application_ids.application_id}/devices/{end_device_ids.device_id}/up/simulate` as an endpoint, and create a loop with a single simulated message - the stack will simulate the uplink, send it to the webhooks, the webhooks call `SimulateUplink` and the process continues forever. It is even possible to make the AS generate more messages by registering two webhooks to the `SimulateUplink` endpoint - you then start the loop by simulating one message, the fanout creates two calls to `SimulateUplink`, and the cycle continues - you get thousands of inflight messages.

Historically the reason why this process actually ends is because on every RPC we would append a new correlation ID, and the correlation IDs have a maximum `repeated` size in the validation, so at some point the loop ends because the AS has appended the correlation ID too many times and the call fails in the validation phase. While this works, I would prefer to block loops explicitly instead of relying on this emergent behavior.

Reply:
> But then we would have to comment that simulated is set by AS, and it is not expected to be set on incoming simulated uplink messages.

I have added an explicit comment to the `simulated` field. For the record, none of our callers (CLI/Console) have manually set it historically.

> Can't we rely on a correlation ID set by AS?

As I mentioned originally, yes we can but it is ad-hoc. I prefer to explicitly disable re-simulating traffic.

> Otherwise we have to update the simulated field in AS I think.

We do - it was always the case that the RPC marks the traffic:
https://github.com/TheThingsNetwork/lorawan-stack/blob/8c54b90132e39406d95b737808aa2d5d21979fda/pkg/applicationserver/io/grpc/grpc.go#L238-L241

> For clarity, we should probably return InvalidArgument instead of silently ignoring the message.

I agree this is better - I have added an error.

Reply:
> > Otherwise we have to update the simulated field in AS I think.
> 
> We do - it was always the case that the RPC marks the traffic:

I meant in the API doc, like you did.